File: uzum/python/uzum_payment/uzum_orders.py
"""
===============================================================================
  SIZ FAQAT SHU FAYLNI TAHRIRLAYSIZ
===============================================================================

Bu fayl Uzum Bank integratsiyasini SIZNING bazangiz bilan bog'laydi.
Qolgan fayllarga (uzum_methods.py, uzum_auth.py, uzum_config.py,
uzum_errors.py) tegish shart emas — ular o'zgarmaydi.

Uzum Bank Payme'ga o'xshaydi: bitta "hisob" (sizning buyurtmangiz) va bitta
"tranzaksiya" (Uzum Bank ochadigan yozuv) bor. Farqi — holatlar soni uchta
(Payme'da to'rtta: 1/2/-1/-2), timeout 30 daqiqa (Payme'da 12 soat), va eng
muhimi: TAKRORIY /create yoki /confirm so'rovida Uzum Bank aniq XATO kutadi
(idempotent muvaffaqiyat emas) — bu quyidagi funksiyalarda hisobga olingan.

Shuning uchun bu faylda IKKI GURUH funksiya bor:

    HISOBINGIZGA ULANISH (siz yozasiz):
      1. find_account(params)         -> buyurtmangizni toping
      2. on_confirmed(transaction)    -> to'lov o'tgach mahsulotni bering
      3. on_reversed(transaction)     -> bekor qilinganda (ixtiyoriy)
      4. can_reverse(transaction)     -> tasdiqlangandan keyin bekor qilish
                                          mumkinmi (ixtiyoriy, standart: ha)

    TRANZAKSIYA KUNDALIGI (Uzum Bank talab qiladi, demo tayyor turibdi):
      Har bir tranzaksiyani vaqti va holati bilan saqlab turish — Uzum
      Bank'ning protokol talabi (/status shuni so'raydi). Hozir SQLite
      namunasi turibdi, production'da almashtirasiz (fayl oxiriga qarang).
===============================================================================
"""

from __future__ import annotations

import logging
import os
import sqlite3
import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

logger = logging.getLogger("uzum")

# --- Tranzaksiya holatlari -----------------------------------------------
# Bu qiymatlarni Uzum Bank belgilagan (status maydonida aynan shu satrlar
# ishlatiladi) — o'zgartirmang.

STATE_CREATED = "CREATED"      # yaratilgan, hali tasdiqlanmagan
STATE_CONFIRMED = "CONFIRMED"  # to'langan (mahsulot berilgan)
STATE_REVERSED = "REVERSED"    # bekor qilingan (to'lanmasdan turib YOKI qaytarilgan)

# 30 daqiqa — shuncha vaqt ichida tasdiqlanmagan (/confirm kelmagan)
# tranzaksiya "muvaffaqiyatsiz" hisoblanadi. Bu — Uzum Bank hujjatida aniq
# yozilgan qoida: BIZ o'zimiz shu muddatni kuzatishimiz kerak (Uzum Bank
# alohida "bekor qilish" so'rovi yubormaydi).
TRANSACTION_TIMEOUT_MS = 30 * 60 * 1000


@dataclass
class UzumAccount:
    """`find_account()` qaytaradigan narsa — sizning buyurtmangiz haqida.

    id
        Bazangizdagi buyurtma id'si.

    amount
        Kutilayotgan summa TIYINDA (1 so'm = 100 tiyin).

    payable
        Buyurtma hali to'lov kutyaptimi? Allaqachon to'langan/bekor
        qilingan bo'lsa `False` qiling.

    extra
        `on_confirmed()` ichida kerak bo'ladigan hamma narsa.
    """

    id: int | str
    amount: int
    payable: bool = True
    extra: dict[str, Any] = field(default_factory=dict)


@dataclass
class UzumTransaction:
    """Uzum Bank tranzaksiya yozuvi — protokol talab qiladigan kundalik yozuvi."""

    trans_id: str                 # Uzum Bank bergan UUID
    params: dict[str, Any]        # foydalanuvchi kiritgan hisob maydonlari
    amount: int
    state: str
    create_time: int
    confirm_time: int = 0
    reverse_time: int = 0
    our_id: str = ""
    account_extra: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not self.our_id:
            self.our_id = self.trans_id


# =============================================================================
#   HISOBINGIZGA ULANISH — shu 4 ta funksiyani o'zgartirasiz
# =============================================================================


def find_account(params: dict[str, Any]) -> UzumAccount | None:
    """Uzum Bank yuborgan `params` maydonlari bo'yicha buyurtmani topadi.

    `params` — foydalanuvchi Uzum Bank ilovasida kiritgan qiymatlar,
    masalan `{"account": "42"}`. Aniq qaysi maydon ishlatilishi Uzum Bank
    kabinetida xizmatingizni sozlaganda siz belgilaysiz.

    Topilmasa `None` qaytaring — Uzum Bank "10007" oladi.
    """
    order_id = params.get("account")
    if order_id is None:
        return None

    row = _demo_db().execute(
        "SELECT * FROM demo_orders WHERE id = ? LIMIT 1", (str(order_id),)
    ).fetchone()

    if row is None:
        return None

    return UzumAccount(
        id=row["id"],
        amount=int(row["amount_tiyin"]),
        payable=(row["status"] == "pending"),
        extra={"product": row["product"]},
    )


def on_confirmed(transaction: UzumTransaction) -> None:
    """To'lov tasdiqlangach BIR MARTA chaqiriladi. Mahsulotni shu yerda bering.

    DIQQAT: Uzum Bank oqimida pul /confirm kelguncha ALLAQACHON yechilgan
    bo'ladi (Uzum foydalanuvchidan pulni /create dan keyin, /confirm dan
    OLDIN yechadi). Shuning uchun bu funksiya xato bersa ham, pul qaytarib
    berilmaydi — faqat mahsulot berish jarayoni chalasi qoladi. Xatoni
    logdan kuzatib, qo'lda hal qilasiz (yoki keyinroq /reverse orqali
    qaytarasiz).
    """
    order_id = transaction.params.get("account")
    with _demo_write() as conn:
        conn.execute(
            "UPDATE demo_orders SET status = 'paid' WHERE id = ?", (str(order_id),)
        )
    logger.info(
        "TASDIQLANDI: order=%s, summa=%s tiyin, trans_id=%s",
        order_id,
        transaction.amount,
        transaction.trans_id,
    )


def on_reversed(transaction: UzumTransaction) -> None:
    """Tranzaksiya bekor qilinganda (yoki tasdiqlangandan keyin qaytarilganda)
    chaqiriladi.

    Agar bekor qilish CONFIRMED holatidan bo'lsa — bu QAYTARISH (pul
    allaqachon berilgan edi). Shu holatda mahsulotga ruxsatni bekor qiling.
    """
    order_id = transaction.params.get("account")
    with _demo_write() as conn:
        conn.execute(
            "UPDATE demo_orders SET status = 'cancelled' WHERE id = ?", (str(order_id),)
        )
    logger.info("BEKOR QILINDI: order=%s, trans_id=%s", order_id, transaction.trans_id)


def can_reverse(transaction: UzumTransaction) -> bool:
    """Tasdiqlangan (CONFIRMED) tranzaksiyani bekor qilish (qaytarish) mumkinmi?

    Standart: har doim mumkin. Mahsulot qaytarib bo'lmaydigan bo'lsa,
    `False` qaytaring — Uzum Bank "10017" oladi.
    """
    return True


# =============================================================================
#   TRANZAKSIYA KUNDALIGI — Uzum Bank talab qiladi, demo tayyor.
#   O'z bazangizga o'tganingizda pastdagi funksiyalarni almashtiring
#   (namunalar fayl oxirida).
# =============================================================================

_conn: sqlite3.Connection | None = None
_lock = threading.Lock()

SCHEMA = """
CREATE TABLE IF NOT EXISTS uzum_transactions (
    trans_id      TEXT    PRIMARY KEY,
    our_id        TEXT    NOT NULL,
    params_json   TEXT    NOT NULL,
    amount        INTEGER NOT NULL,
    state         TEXT    NOT NULL,
    create_time   INTEGER NOT NULL,
    confirm_time  INTEGER NOT NULL DEFAULT 0,
    reverse_time  INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_uzum_tx_create_time ON uzum_transactions(create_time);

CREATE TABLE IF NOT EXISTS demo_orders (
    id            TEXT    PRIMARY KEY,
    amount_tiyin  INTEGER NOT NULL,
    status        TEXT    NOT NULL DEFAULT 'pending',
    product       TEXT
);
"""


def _demo_db() -> sqlite3.Connection:
    global _conn
    if _conn is None:
        path = os.getenv("UZUM_DB_PATH", "uzum_demo.db")
        _conn = sqlite3.connect(path, check_same_thread=False)
        _conn.row_factory = sqlite3.Row
        if path != ":memory:":
            _conn.execute("PRAGMA journal_mode=WAL")
        _conn.executescript(SCHEMA)
        _conn.commit()
    return _conn


class _demo_write:
    """Yozish tranzaksiyasi (bir vaqtda bitta yozuvchi)."""

    def __enter__(self) -> sqlite3.Connection:
        _lock.acquire()
        self.conn = _demo_db()
        self.conn.execute("BEGIN IMMEDIATE")
        return self.conn

    def __exit__(self, exc_type, exc, tb) -> None:
        try:
            if exc_type is None:
                self.conn.commit()
            else:
                self.conn.rollback()
        finally:
            _lock.release()


def _row_to_transaction(row: sqlite3.Row) -> UzumTransaction:
    import json

    return UzumTransaction(
        trans_id=row["trans_id"],
        our_id=row["our_id"],
        params=json.loads(row["params_json"]),
        amount=int(row["amount"]),
        state=row["state"],
        create_time=int(row["create_time"]),
        confirm_time=int(row["confirm_time"]),
        reverse_time=int(row["reverse_time"]),
    )


def get_transaction(trans_id: str) -> UzumTransaction | None:
    """Uzum Bank'ning tranzaksiya id'si (`transId`) bo'yicha yozuvni topadi."""
    row = _demo_db().execute(
        "SELECT * FROM uzum_transactions WHERE trans_id = ? LIMIT 1", (trans_id,)
    ).fetchone()
    return _row_to_transaction(row) if row else None


def get_active_transaction_for_account(params: dict[str, Any]) -> UzumTransaction | None:
    """Shu hisob uchun hali bekor qilinmagan tranzaksiya bormi?

    /create bitta buyurtmaga ikkita PARALLEL faol tranzaksiya ochilishining
    oldini olish uchun ishlatadi.
    """
    import json

    order_id = params.get("account")
    rows = _demo_db().execute(
        "SELECT * FROM uzum_transactions WHERE state IN (?, ?)",
        (STATE_CREATED, STATE_CONFIRMED),
    ).fetchall()
    for row in rows:
        p = json.loads(row["params_json"])
        if p.get("account") == order_id:
            return _row_to_transaction(row)
    return None


def create_transaction(trans_id: str, amount: int, params: dict[str, Any]) -> UzumTransaction:
    """Yangi tranzaksiya yozuvini yaratadi (state=CREATED)."""
    import json

    now = int(datetime.now(timezone.utc).timestamp() * 1000)

    with _demo_write() as conn:
        conn.execute(
            """INSERT INTO uzum_transactions
                   (trans_id, our_id, params_json, amount, state, create_time)
               VALUES (?, ?, ?, ?, ?, ?)""",
            (trans_id, trans_id, json.dumps(params), amount, STATE_CREATED, now),
        )

    return UzumTransaction(
        trans_id=trans_id,
        our_id=trans_id,
        params=params,
        amount=amount,
        state=STATE_CREATED,
        create_time=now,
    )


def mark_confirmed(trans_id: str) -> UzumTransaction | None:
    """Tranzaksiyani "tasdiqlangan" qiladi (state: CREATED -> CONFIRMED).

    Faqat HAQIQATAN o'tkazgan chaqiruv yangilangan yozuvni qaytaradi — aks
    holda `None`. Bu — takroriy so'rovda `on_confirmed()` qayta
    chaqirilmasligi uchun MUHIM (garchi Uzum Bank protokoli bo'yicha
    takroriy /confirm asosan xato bilan rad etilsa ham, himoya ortiqcha
    emas — masalan parallel so'rov holatida).
    """
    now = int(datetime.now(timezone.utc).timestamp() * 1000)
    with _demo_write() as conn:
        cur = conn.execute(
            """UPDATE uzum_transactions
               SET state = ?, confirm_time = ?
               WHERE trans_id = ? AND state = ?""",
            (STATE_CONFIRMED, now, trans_id, STATE_CREATED),
        )
        if cur.rowcount == 0:
            return None
    return get_transaction(trans_id)


def mark_reversed(trans_id: str) -> UzumTransaction | None:
    """Tranzaksiyani bekor qiladi (CREATED yoki CONFIRMED -> REVERSED).

    Allaqachon REVERSED bo'lsa `None` qaytaradi.
    """
    now = int(datetime.now(timezone.utc).timestamp() * 1000)

    current = get_transaction(trans_id)
    if current is None or current.state == STATE_REVERSED:
        return None

    with _demo_write() as conn:
        conn.execute(
            """UPDATE uzum_transactions
               SET state = ?, reverse_time = ?
               WHERE trans_id = ?""",
            (STATE_REVERSED, now, trans_id),
        )

    return get_transaction(trans_id)


def demo_create_order(order_id: str, amount_tiyin: int, product: str = "") -> None:
    """Namuna uchun buyurtma yaratadi (demo_orders jadvaliga)."""
    with _demo_write() as conn:
        conn.execute(
            "INSERT OR REPLACE INTO demo_orders (id, amount_tiyin, status, product) "
            "VALUES (?, ?, 'pending', ?)",
            (order_id, amount_tiyin, product),
        )


def reset_db_for_tests() -> None:
    """Testlar uchun ulanishni tozalaydi."""
    global _conn
    if _conn is not None:
        _conn.close()
    _conn = None


# =============================================================================
#   O'Z BAZANGIZ UCHUN NAMUNA — nusxa oling va yuqoridagi tranzaksiya
#   kundaligi funksiyalari o'rniga qo'ying (PostgreSQL, psycopg misolida).
# =============================================================================
#
#   import psycopg
#   from psycopg.rows import dict_row
#
#   def _pg():
#       return psycopg.connect(os.environ["DATABASE_URL"], row_factory=dict_row)
#
#   def get_transaction(trans_id):
#       with _pg() as c, c.cursor() as cur:
#           cur.execute("SELECT * FROM uzum_transactions WHERE trans_id=%s", (trans_id,))
#           row = cur.fetchone()
#       if not row:
#           return None
#       return UzumTransaction(
#           trans_id=row["trans_id"], our_id=row["our_id"], params=row["params_json"],
#           amount=row["amount"], state=row["state"], create_time=row["create_time"],
#           confirm_time=row["confirm_time"], reverse_time=row["reverse_time"],
#       )
#
#   def mark_confirmed(trans_id):
#       with _pg() as c, c.cursor() as cur:
#           cur.execute(
#               "UPDATE uzum_transactions SET state='CONFIRMED', confirm_time=%s "
#               "WHERE trans_id=%s AND state='CREATED'",
#               (now_ms(), trans_id),
#           )
#           if cur.rowcount == 0:      # <- atomar, poyga yo'q
#               return None
#       return get_transaction(trans_id)
#
#   find_account(), on_confirmed(), on_reversed() — bularni O'ZGARTIRMAYSIZ
#   (ular ledger emas, sizning biznesingiz) — faqat ichlarida o'z
#   ORM/SQL'ingizni ishlatasiz (Click/Payme'dagi bilan bir xil naqsh).
# =============================================================================

File: uzum/php/uzum_orders.php
<?php
/**
 * =============================================================================
 *   SIZ FAQAT SHU FAYLNI TAHRIRLAYSIZ
 * =============================================================================
 *
 * Bu fayl Uzum Bank integratsiyasini SIZNING bazangiz bilan bog'laydi.
 * Qolgan fayllarga (uzum_methods.php, uzum_auth.php, uzum_config.php,
 * uzum_errors.php) tegish shart emas — ular o'zgarmaydi.
 *
 * Uzum Bank Payme'ga o'xshaydi: bitta "hisob" (sizning buyurtmangiz) va
 * bitta "tranzaksiya" (Uzum Bank ochadigan yozuv) bor. Farqi — holatlar
 * uchta (Payme'da to'rtta), timeout 30 daqiqa (Payme'da 12 soat), va eng
 * muhimi: TAKRORIY /create yoki /confirm so'rovida Uzum Bank aniq XATO
 * kutadi (idempotent muvaffaqiyat emas).
 *
 * Shuning uchun bu faylda IKKI GURUH funksiya bor:
 *
 *     HISOBINGIZGA ULANISH (siz yozasiz):
 *       1. uzumFindAccount($params)         -> buyurtmangizni toping
 *       2. uzumOnConfirmed($transaction)     -> to'lov o'tgach mahsulotni bering
 *       3. uzumOnReversed($transaction)      -> bekor qilinganda (ixtiyoriy)
 *       4. uzumCanReverse($transaction)      -> tasdiqlangandan keyin bekor
 *                                                qilish mumkinmi (ixtiyoriy)
 *
 *     TRANZAKSIYA KUNDALIGI (Uzum Bank talab qiladi, demo tayyor turibdi):
 *       Har bir tranzaksiyani vaqti va holati bilan saqlab turish — Uzum
 *       Bank'ning protokol talabi (/status shuni so'raydi). Hozir SQLite
 *       namunasi turibdi, production'da almashtirasiz (fayl oxiriga qarang).
 * =============================================================================
 */

require_once __DIR__ . '/uzum_config.php';

// --- Holatlar ------------------------------------------------------------
// Bu qiymatlarni Uzum Bank belgilagan (status maydonida aynan shu satrlar
// ishlatiladi) — o'zgartirmang.

define('UZUM_STATE_CREATED', 'CREATED');
define('UZUM_STATE_CONFIRMED', 'CONFIRMED');
define('UZUM_STATE_REVERSED', 'REVERSED');

// 30 daqiqa — shuncha vaqt ichida tasdiqlanmagan tranzaksiya "muvaffaqiyatsiz"
// hisoblanadi. Buni Uzum Bank alohida xabar bermaydi — o'zimiz kuzatamiz.
define('UZUM_TRANSACTION_TIMEOUT_MS', 30 * 60 * 1000);

/**
 * `uzumFindAccount()` qaytaradigan narsa — sizning buyurtmangiz haqida.
 */
class UzumAccount
{
    /** @var int|string */
    public $id;
    /** @var int summa TIYINDA */
    public $amount;
    /** @var bool */
    public $payable;
    /** @var array */
    public $extra;

    public function __construct($id, $amount, $payable = true, array $extra = array())
    {
        $this->id = $id;
        $this->amount = (int)$amount;
        $this->payable = (bool)$payable;
        $this->extra = $extra;
    }
}

/**
 * Uzum Bank tranzaksiya yozuvi — protokol talab qiladigan kundalik yozuvi.
 */
class UzumTransaction
{
    /** @var string Uzum Bank bergan UUID */
    public $transId;
    /** @var array foydalanuvchi kiritgan hisob maydonlari */
    public $params;
    /** @var int */
    public $amount;
    /** @var string CREATED|CONFIRMED|REVERSED */
    public $state;
    /** @var int */
    public $createTime;
    /** @var int */
    public $confirmTime = 0;
    /** @var int */
    public $reverseTime = 0;
    /** @var string */
    public $ourId;
    /** @var array */
    public $accountExtra = array();

    public function __construct($transId, array $params, $amount, $state, $createTime)
    {
        $this->transId = $transId;
        $this->ourId = $transId;
        $this->params = $params;
        $this->amount = (int)$amount;
        $this->state = $state;
        $this->createTime = (int)$createTime;
    }
}

// =============================================================================
//   HISOBINGIZGA ULANISH — shu 4 ta funksiyani o'zgartirasiz
// =============================================================================

if (!function_exists('uzumFindAccount')) {
    /**
     * Uzum Bank yuborgan `params` maydonlari bo'yicha buyurtmani topadi.
     *
     * `params` — foydalanuvchi Uzum Bank ilovasida kiritgan qiymatlar,
     * masalan `array('account' => '42')`.
     *
     * Topilmasa `null` qaytaring — Uzum Bank "10007" oladi.
     *
     * @return UzumAccount|null
     */
    function uzumFindAccount(array $params)
    {
        if (!isset($params['account'])) {
            return null;
        }
        $orderId = (string)$params['account'];

        $stmt = uzumDemoDb()->prepare('SELECT * FROM demo_orders WHERE id = ? LIMIT 1');
        $stmt->execute(array($orderId));
        $row = $stmt->fetch();

        if (!$row) {
            return null;
        }

        return new UzumAccount(
            $row['id'],
            $row['amount_tiyin'],
            $row['status'] === 'pending',
            array('product' => $row['product'])
        );
    }
}

if (!function_exists('uzumOnConfirmed')) {
    /**
     * To'lov tasdiqlangach BIR MARTA chaqiriladi. Mahsulotni shu yerda bering.
     *
     * DIQQAT: Uzum Bank oqimida pul /confirm kelguncha ALLAQACHON
     * yechilgan bo'ladi. Shuning uchun bu funksiya xato bersa ham, pul
     * qaytarib berilmaydi — faqat mahsulot berish jarayoni chalasi qoladi.
     * Xatoni logdan kuzatib, qo'lda hal qilasiz.
     */
    function uzumOnConfirmed(UzumTransaction $transaction)
    {
        $orderId = isset($transaction->params['account']) ? $transaction->params['account'] : null;

        $stmt = uzumDemoDb()->prepare("UPDATE demo_orders SET status = 'paid' WHERE id = ?");
        $stmt->execute(array((string)$orderId));

        uzumLog('info', 'TASDIQLANDI', array(
            'order_id' => $orderId,
            'amount'   => $transaction->amount,
            'trans_id' => $transaction->transId,
        ));
    }
}

if (!function_exists('uzumOnReversed')) {
    /**
     * Tranzaksiya bekor qilinganda (yoki qaytarilganda) chaqiriladi.
     */
    function uzumOnReversed(UzumTransaction $transaction)
    {
        $orderId = isset($transaction->params['account']) ? $transaction->params['account'] : null;

        $stmt = uzumDemoDb()->prepare("UPDATE demo_orders SET status = 'cancelled' WHERE id = ?");
        $stmt->execute(array((string)$orderId));

        uzumLog('info', 'BEKOR QILINDI', array('order_id' => $orderId, 'trans_id' => $transaction->transId));
    }
}

if (!function_exists('uzumCanReverse')) {
    /**
     * Tasdiqlangan (CONFIRMED) tranzaksiyani bekor qilish (qaytarish)
     * mumkinmi? Standart: har doim mumkin.
     */
    function uzumCanReverse(UzumTransaction $transaction)
    {
        return true;
    }
}

if (!function_exists('uzumLog')) {
    function uzumLog($level, $message, array $context = array())
    {
        $line = '[uzum] [' . strtoupper($level) . '] ' . $message;
        if (!empty($context)) {
            $line .= ' ' . json_encode($context, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
        }
        error_log($line);
    }
}

// =============================================================================
//   TRANZAKSIYA KUNDALIGI — Uzum Bank talab qiladi, demo tayyor.
//   O'z bazangizga o'tganingizda pastdagi funksiyalarni almashtiring
//   (namunalar fayl oxirida).
// =============================================================================

if (!function_exists('uzumDemoDb')) {
    function uzumDemoDb()
    {
        static $pdo = null;

        if ($pdo === null) {
            $path = uzumEnv('UZUM_DB_PATH', __DIR__ . '/uzum_demo.sqlite');

            $pdo = new PDO('sqlite:' . $path);
            $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
            $pdo->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);

            $pdo->exec(
                "CREATE TABLE IF NOT EXISTS uzum_transactions (
                    trans_id     TEXT PRIMARY KEY,
                    our_id       TEXT NOT NULL,
                    params_json  TEXT NOT NULL,
                    amount       INTEGER NOT NULL,
                    state        TEXT NOT NULL,
                    create_time  INTEGER NOT NULL,
                    confirm_time INTEGER NOT NULL DEFAULT 0,
                    reverse_time INTEGER NOT NULL DEFAULT 0
                )"
            );
            $pdo->exec('CREATE INDEX IF NOT EXISTS idx_uzum_tx_create_time ON uzum_transactions(create_time)');
            $pdo->exec(
                "CREATE TABLE IF NOT EXISTS demo_orders (
                    id           TEXT PRIMARY KEY,
                    amount_tiyin INTEGER NOT NULL,
                    status       TEXT NOT NULL DEFAULT 'pending',
                    product      TEXT
                )"
            );
        }

        return $pdo;
    }
}

if (!function_exists('uzumNowMs')) {
    function uzumNowMs()
    {
        return (int)round(microtime(true) * 1000);
    }
}

if (!function_exists('uzumRowToTransaction')) {
    function uzumRowToTransaction(array $row)
    {
        $tx = new UzumTransaction(
            $row['trans_id'],
            json_decode($row['params_json'], true),
            $row['amount'],
            $row['state'],
            $row['create_time']
        );
        $tx->ourId = $row['our_id'];
        $tx->confirmTime = (int)$row['confirm_time'];
        $tx->reverseTime = (int)$row['reverse_time'];

        return $tx;
    }
}

if (!function_exists('uzumGetTransaction')) {
    /**
     * Uzum Bank'ning tranzaksiya id'si (`transId`) bo'yicha yozuvni topadi.
     *
     * @return UzumTransaction|null
     */
    function uzumGetTransaction($transId)
    {
        $stmt = uzumDemoDb()->prepare('SELECT * FROM uzum_transactions WHERE trans_id = ? LIMIT 1');
        $stmt->execute(array($transId));
        $row = $stmt->fetch();

        return $row ? uzumRowToTransaction($row) : null;
    }
}

if (!function_exists('uzumGetActiveTransactionForAccount')) {
    /**
     * Shu hisob uchun hali bekor qilinmagan tranzaksiya bormi?
     *
     * /create bitta buyurtmaga ikkita PARALLEL faol tranzaksiya
     * ochilishining oldini olish uchun ishlatadi.
     *
     * @return UzumTransaction|null
     */
    function uzumGetActiveTransactionForAccount(array $params)
    {
        $orderId = isset($params['account']) ? $params['account'] : null;

        $stmt = uzumDemoDb()->prepare(
            'SELECT * FROM uzum_transactions WHERE state IN (?, ?)'
        );
        $stmt->execute(array(UZUM_STATE_CREATED, UZUM_STATE_CONFIRMED));

        while ($row = $stmt->fetch()) {
            $p = json_decode($row['params_json'], true);
            $pAccount = isset($p['account']) ? $p['account'] : null;
            if ($pAccount == $orderId) {
                return uzumRowToTransaction($row);
            }
        }

        return null;
    }
}

if (!function_exists('uzumCreateTransaction')) {
    /**
     * Yangi tranzaksiya yozuvini yaratadi (state=CREATED).
     *
     * @return UzumTransaction
     */
    function uzumCreateTransaction($transId, $amount, array $params)
    {
        $now = uzumNowMs();

        $stmt = uzumDemoDb()->prepare(
            'INSERT INTO uzum_transactions
                 (trans_id, our_id, params_json, amount, state, create_time)
             VALUES (?, ?, ?, ?, ?, ?)'
        );
        $stmt->execute(array(
            $transId,
            $transId,
            json_encode($params),
            (int)$amount,
            UZUM_STATE_CREATED,
            $now,
        ));

        return new UzumTransaction($transId, $params, $amount, UZUM_STATE_CREATED, $now);
    }
}

if (!function_exists('uzumMarkConfirmed')) {
    /**
     * Tranzaksiyani "tasdiqlangan" qiladi (state: CREATED -> CONFIRMED).
     *
     * Faqat HAQIQATAN o'tkazgan chaqiruv yangilangan yozuvni qaytaradi —
     * aks holda `null`.
     *
     * @return UzumTransaction|null
     */
    function uzumMarkConfirmed($transId)
    {
        $now = uzumNowMs();

        $stmt = uzumDemoDb()->prepare(
            "UPDATE uzum_transactions
                SET state = ?, confirm_time = ?
              WHERE trans_id = ? AND state = ?"
        );
        $stmt->execute(array(UZUM_STATE_CONFIRMED, $now, $transId, UZUM_STATE_CREATED));

        if ($stmt->rowCount() === 0) {
            return null;
        }

        return uzumGetTransaction($transId);
    }
}

if (!function_exists('uzumMarkReversed')) {
    /**
     * Tranzaksiyani bekor qiladi (CREATED yoki CONFIRMED -> REVERSED).
     *
     * Allaqachon REVERSED bo'lsa `null` qaytaradi.
     *
     * @return UzumTransaction|null
     */
    function uzumMarkReversed($transId)
    {
        $now = uzumNowMs();

        $current = uzumGetTransaction($transId);
        if ($current === null || $current->state === UZUM_STATE_REVERSED) {
            return null;
        }

        $stmt = uzumDemoDb()->prepare(
            "UPDATE uzum_transactions SET state = ?, reverse_time = ? WHERE trans_id = ?"
        );
        $stmt->execute(array(UZUM_STATE_REVERSED, $now, $transId));

        return uzumGetTransaction($transId);
    }
}

if (!function_exists('uzumDemoCreateOrder')) {
    /**
     * Namuna uchun buyurtma yaratadi (demo_orders jadvaliga).
     */
    function uzumDemoCreateOrder($orderId, $amountTiyin, $product = '')
    {
        $stmt = uzumDemoDb()->prepare(
            "INSERT OR REPLACE INTO demo_orders (id, amount_tiyin, status, product)
             VALUES (?, ?, 'pending', ?)"
        );
        $stmt->execute(array((string)$orderId, (int)$amountTiyin, $product));
    }
}

// =============================================================================
//   O'Z BAZANGIZ UCHUN NAMUNA — nusxa oling va yuqoridagi tranzaksiya
//   kundaligi funksiyalari o'rniga qo'ying (MySQL misolida).
// =============================================================================
//
//   function uzumGetTransaction($transId)
//   {
//       $stmt = db()->prepare('SELECT * FROM uzum_transactions WHERE trans_id = ?');
//       $stmt->execute(array($transId));
//       $row = $stmt->fetch();
//       if (!$row) return null;
//       // ... uzumRowToTransaction($row) bilan bir xil qaytaring
//   }
//
//   function uzumMarkConfirmed($transId)
//   {
//       $stmt = db()->prepare(
//           "UPDATE uzum_transactions SET state='CONFIRMED', confirm_time=?
//             WHERE trans_id=? AND state='CREATED'"
//       );
//       $stmt->execute(array(uzumNowMs(), $transId));
//       if ($stmt->rowCount() === 0) {    // <- atomar, poyga yo'q
//           return null;
//       }
//       return uzumGetTransaction($transId);
//   }
//
//   uzumFindAccount(), uzumOnConfirmed(), uzumOnReversed() — bularni
//   O'ZGARTIRMAYSIZ (ular ledger emas, sizning biznesingiz) — faqat
//   ichlarida o'z SQL'ingizni ishlatasiz.
// =============================================================================
